From 0bb3191fe16eb087502693b2864b88898a0be764 Mon Sep 17 00:00:00 2001
From: Takeshi Kihara <takeshi.kihara.df@renesas.com>
Date: Mon, 28 Jan 2019 09:03:43 +0900
Subject: [PATCH 726/909] clk: renesas: r8a7796: Remove iVDP1C and FCPCI0
 clocks on ES3.0

commit 59fcadc987f6bac8f70e330294c6df6dce156c0b from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

Module Standby and Software Reset specifications differ between R-Car M3
ES1.x to ES3.0 or later:
  - From ES3.0, the FCPCI0 module clock bit of the MSTPSR6/RMSTPCR6/
    SMSTPCR6/SRCR6 register was removed.
  - From ES3.0, the iVDP1C module clock bit of the MSTPSR1/RMSTPCR1/
    SMSTPCR1/SRCR1 register was merged from bit28 to bit30 which
    the VCP4(VCPLF) module clock bit.

This patch removes the FCPCI0 and iVDP1C module clocks. To control
the iVDP1C module clock, the VCP4(VCPLF) module clock bit is used.

Signed-off-by: Takeshi Kihara <takeshi.kihara.df@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/clk/renesas/r8a7796-cpg-mssr.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/renesas/r8a7796-cpg-mssr.c b/drivers/clk/renesas/r8a7796-cpg-mssr.c
index 9064657ca605..378b34442974 100644
--- a/drivers/clk/renesas/r8a7796-cpg-mssr.c
+++ b/drivers/clk/renesas/r8a7796-cpg-mssr.c
@@ -2,7 +2,7 @@
  * r8a7796 Clock Pulse Generator / Module Standby and Software Reset
  *
  * Copyright (C) 2016 Glider bvba
- * Copyright (C) 2018 Renesas Electronics Corp.
+ * Copyright (C) 2018-2019 Renesas Electronics Corp.
  *
  * Based on r8a7795-cpg-mssr.c
  *
@@ -18,6 +18,7 @@
 #include <linux/init.h>
 #include <linux/kernel.h>
 #include <linux/soc/renesas/rcar-rst.h>
+#include <linux/sys_soc.h>
 
 #include <dt-bindings/clock/r8a7796-cpg-mssr.h>
 
@@ -119,7 +120,7 @@ static const struct cpg_core_clk r8a7796_core_clks[] __initconst = {
 	DEF_BASE("r",           R8A7796_CLK_R,     CLK_TYPE_GEN3_R, CLK_RINT),
 };
 
-static const struct mssr_mod_clk r8a7796_mod_clks[] __initconst = {
+static struct mssr_mod_clk r8a7796_mod_clks[] __initdata = {
 	DEF_MOD("3dge",			 112,	R8A7796_CLK_ZG),
 	DEF_MOD("fdp1-0",		 119,	R8A7796_CLK_S0D1),
 	DEF_MOD("ivdp1c",		 128,	R8A7796_CLK_S0D2),
@@ -302,6 +303,20 @@ static const struct rcar_gen3_cpg_pll_config cpg_pll_configs[16] __initconst = {
 	{ 2,		192,	1,	192,	1,	32,	},
 };
 
+static const struct soc_device_attribute r8a7796es1[] __initconst = {
+	{ .soc_id = "r8a7796", .revision = "ES1.*" },
+	{ /* sentinel */ }
+};
+
+	/*
+	 * Fixups for R-Car M3 ES3.x
+	 */
+
+static const unsigned int r8a7796es3_mod_nullify[] __initconst = {
+	MOD_CLK_ID(128),			/* iVDP1C  */
+	MOD_CLK_ID(617),			/* FCPCI0  */
+};
+
 static int __init r8a7796_cpg_mssr_init(struct device *dev)
 {
 	const struct rcar_gen3_cpg_pll_config *cpg_pll_config;
@@ -312,6 +327,12 @@ static int __init r8a7796_cpg_mssr_init(struct device *dev)
 	if (error)
 		return error;
 
+	if (!soc_device_match(r8a7796es1))
+		mssr_mod_nullify(r8a7796_mod_clks,
+				 ARRAY_SIZE(r8a7796_mod_clks),
+				 r8a7796es3_mod_nullify,
+				 ARRAY_SIZE(r8a7796es3_mod_nullify));
+
 	cpg_pll_config = &cpg_pll_configs[CPG_PLL_CONFIG_INDEX(cpg_mode)];
 	if (!cpg_pll_config->extal_div) {
 		dev_err(dev, "Prohibited setting (cpg_mode=0x%x)\n", cpg_mode);
-- 
2.17.1

